src/OpenTelemetry.Api/Trace/TracerProvider.cs

```diff
@@ -44,5 +51,92 @@ protected TracerProvider()
     /// <param name="version">Version of the instrumentation library.</param>
     /// <returns>Tracer instance.</returns>
     public Tracer GetTracer(string name, string? version = null)
-        => new(new ActivitySource(name ?? string.Empty, version));
+    {
+        var tracers = this.tracers;
+        if (tracers == null)
+        {
+            // Note: Returns a no-op Tracer once dispose has been called.
+            return new(activitySource: null);
+        }
+
+        var key = threadTracerKey == null
+            ? (threadTracerKey = new TracerKey(name, version))
+            : threadTracerKey.Update(name, version);
+
+        if (tracers[key] is not Tracer tracer)
+        {
+            lock (tracers)
+            {
+                tracer = (tracers[key] as Tracer)!;
+                if (tracer == null)
+                {
+                    if (this.tracers != null)
+                    {
+                        tracer = new(new(key.Name, key.Version));
+
+                        // Note: key is copied if we write into the hashtable
+                        // because it must hold onto something immutable.
+                        tracers[key.Copy()] = tracer;
+                    }
+                }
+            }
+        }
+
+        return tracer ?? new(activitySource: null);
+    }
+
+    /// <inheritdoc/>
+    protected override void Dispose(bool disposing)
+    {
+        if (disposing)
+        {
+            var tracers = Interlocked.CompareExchange(ref this.tracers, null, this.tracers);
```

Comment:
Set `this.tracers` to `null` inside the lock? A thread calling `GetTracer` might enter this if condition and then another thread calling `Dispose` could set `tracers` to `null` before the other thread calls `tracers[key.Copy()] = tracer;`.

```csharp
if (this.tracers != null)
{
    tracer = new(new(key.Name, key.Version));

     // Note: key is copied if we write into the hashtable
     // because it must hold onto something immutable.
     tracers[key.Copy()] = tracer;
}
```

Reply:
See above.